### ISSUE TITLE
Callbacks : on_test_begin and on_test_end

### ISSUE THREAD

**Issue by sebastienwood** [fastai/fastai#2120]
Hi,

on the implementation/documentation of the callbacks, there is one feature which is unclear for newcomers (I'm still unsure after navigating the source). There is the flag train to know if we're in train mode or not, but there aren't eg test related callbacks. More specifically, there is no information on the `on_train_begin` and `on_train_end` behavior when we are not in train mode. Would they be executed then ?

Use case for exemple is Binary Connect (Courbariaux et al. 2015) : the network is in FP32 for any operation except forward pass at training and test time. We could get the model parameters and tinker with it to ensure the stuff is binarized for the test but it would then mean code duplication between the Fastai callback at training time and the custom test operations.

(I don't know if it's related to the absence of off the shelf test set operations in the library)

**Comment by sgugger** [comment]
I'm not sure what you're asking for: the callback system is designed for the training loop, so on_train_begin and on_train_end are called respectively at the beginning of the training loop and at the end for setup/clean-up.
Model is put on training mode at the beginning of each epoch just before on_epoch_begin, and set on test mode at the beginning of each validation phase.

If you're talking about inference, the callbacks aren't called in get_preds/predict, apart from on_batch_begin.

**Comment by sebastienwood** [comment]
It would be for the test mode of each validation phase : I have some callbacks that would need to be called under a normal test/validation loop (ie no gradient descent). It would then be of use also at inference time.

On_batch_begin would be a good candidate callback as per your reply but then it would miss an on_batch_end to ensure I would put the FP32 weights when going out of inference/test/validation.

**Comment by sgugger** [comment]
For the training loop you should use on_batch_begin and on_batch_end and control the flag train. This would also work with get_preds and predict normally, as those are the callbacks always called.
